Scripting.md

```diff
@@ -0,0 +1,224 @@
+  
+  - Explicit casting between types is necessary: 
+  ```py
+  var a: float = 10.0
+  var b: int = 25
+  var c: float = float(b)+a
+  var d: int = int(a)+b
+  var e = 0xff
+  var f = b + int(e)
+  ```
+
+#### Arrays
+  - Create array: `var arr <- [{auto "one"; "two"; "three"}]`
+  - Push a value: `push(arr,"new value")`
+  - Length of array: `var len = length(arr)`
+  - Erase from array: `erase(arr,2)`
+  - Clear array: `clear(arr)`
+
+#### Tables
+  - Created table: `var tab: table<string,int>`
+  - Write a value: `tab["key"] = 10`.
+  - Read a value: `var value = find(tab,"key")`
+  - Check if a value exists: `var exists = key_exists(tab,"key")`
+
+### Blocks and flow control
+
+Just like in python, empty blocks are not allowed and must at least have a "pass" statement in them if there are no other statements.
+
+#### Functions
+```py
+def my_function(arg1: int, arg2: string): int
+  return 25;
+```
+
+#### For loops
+
+Ranges:
+```py
+for i in range(0,10)
+  pass
+```
+
+#### Arrays
+```py
+var arr <- [{auto "one"; "two"; "three"}]
+for val in arr
+  print("{val}")
+```
+
+#### Tables
+```py
+var tab: table<string,int>
+for k,v in keys(tab), values(tab)
+  print("{k} {v}")
+```
+
+#### While loops
+```py
+while some_value
+  pass
+```
+
+#### If statements
+```py
+if some_value
+  pass
+elif some_other_value
+  pass
+```
+
+**TODO**: Explain structs
+
+### Exporting/Importing
+
+We can import the noggit API with the statement `require noggit` at the top of our script files,
+and similarly we can import other .das files the same way. To allow other modules to access a function, use the same `[export]` tag as on event listeners:
+
+**file_a.das**:
+```py
+def local_function()
+  print("This function is only accessible inside file_a.das")
+
+[export]
+def library_function()
+  print("This function is accessible to any file that 'requires' it")
+```
+
+**file_b.das**:
+
+```py
+require noggit
+require file_a
+
+[export]
+def left_click()
+  print("This is an event listener, so it also has the 'export' tag")
+  library_function()
+```
+
+### Noggit API
+
+The Noggit API are the functions we expose from inside Noggit to scripts. The file `scripts/noggit.spec.das` contains specifications of all the functions we expose.
+`.spec.das` files contain no actual implementations since they only describe existing C++ functions, and we use them similarly to `d.ts` files in JavaScript.
+
+To use the Noggit API, your script files must contain the line `require noggit` at the top. 
+If you can't see any noggit functions in your autocompletions, it's likely you forgot this line.
+
+Noggit API functions should not be called globally in your scripts, but we still made them globally accessible to make them less cumbersome to write.
+The following two rules define where it's safe to call Noggit functions from in your scripts:
+
+- All functions `add_x_param` should be called from the `select` function **only** (or any functions it calls).
```

Comment:
not sure exactly what you mean by scope state, do you mean a context parameter or just a C++-side state that checks the type of function with every function call?

```
[export]
def select(ctx)
    add_x_param(ctx,....)

def left_click(ctx2)
    print("{pos(ctx2)}")
```

or:

```
void add_x_param()
{
    if(!is_calling_from_select()) throw ...
}
```

Reply:
I was referring to the second one, checking on the c++ side. I see that you are doing a check in `get_ctx()` so this might already be the case and perfectly fine as is.

Reply:
Most functions end up using the script context to access the world or some parameter there which will throw an error if it's null, but I could add checks to a few of the functions that don't. (and especially the add_parameter stuff). script_context itself is only valid for click events.